### ISSUE TITLE
`approval-voting`: merge assignments per tranche and batch verify vrf signatures

### ISSUE THREAD

**Issue by sandreim** [paritytech/polkadot#6618]
As pointed out in https://github.com/paritytech/polkadot/issues/6610#issuecomment-1401468099 we can save a lot of CPU and reduce gossip by doing only one RelayVrfModulo signature per authority and derive multiple assignments from the output which stacks with batch verification.

What it is not clear at least for me is if we can also merge Delay assignments per tranche for the same signer by signing once for multiple candidates. Depending on the distribution this can also save us CPU when importing and reduce the network load.

Another question is why would we need authorities to refuse some assignments ? Is this for being able to control the amount of work to be done ?

CC @burdges

**Comment by burdges** [comment]
Importantly, we'll only consolidate the RelayVrfModulo checks this way.  RelayVrfDelay checks should each remain independent, but they should be only maybe 10% of the checks under normal conditions.

**Comment by burdges** [comment]
After verifying the `VRFInOut` in the signature, the output routine maybe looks vaguely like:
```
const MAX_MODULO_SAMPLES: usize = 32;

/// Iterates over all parachains which we check in tranche zero
fn relay_vrf_modulo(io: VRFInOut, num_samples: usize, max_cores: usize) -> impl Iterator<Item=u16> {
    vrf_io.make_bytes::<[u8; 2*MAX_MODULO_SAMPLES]>(b"RelayVrfModuloIO")
    .chunks_exact(2)
    .take(num_samples)
    .map(|x| u16::from_le_bytes( array_ref![x,0,2] ) % max_cores )
}
```

**Comment by rphmeier** [comment]
Isn't it a problem to derive all `RelayVrfDelay` assignments from a single VRF Output, because then publishing one assignment leaks all the tranches for other candidates you're assigned to? `RelayVrfModulo` doesn't have this issue because all assignments are tranche zero. I also don't think that nodes should be allowed to silently "not take" assignments, at least at the moment. If so, then we'd need a way to update the network as to which assignments are being taken.

**Comment by burdges** [comment]
Yes exactly, RelayVrfModulo could all be derived from one output, but the RelayVrfDelay in different tranches must stay independent.

> I also don't think that nodes should be allowed to silently "not take" assignments, at least at the moment.

It's better to reject, or not announce, an assignment than to announce it and then no show.  I've no idea if we'll ever need this, but I made the RelayVrfModulo independent purely to give us this option, which now looks like a mistake.
